### ISSUE TITLE
Re-created file sometime still has old content for other participants

### ISSUE THREAD

**Issue by tobous** [saros-project/saros#780]
When deleting a file that is currently being worked on in a multi-participant session (more than 2 participants) and then re-creating it with a different content, it sometimes retains its old content for some of the participants.

The whole issue seemed very inconsistent during testing. It did not seem consistent whose editor kept the old content and whether this was influenced by the testing setup.

The tested functionality of re-creating a file in a session is only possible with #714.

The tests in which this issue occurred were done before the activity filter introduced in #714 was correctly adjusted for editor activities, meaning it dropped the editor closed activity of the user that processed the deletion first. This may or may not be relevant to this issue.

#### Steps to reproduce
This issue seems to be very inconsistent. We also only got it to work in a three-participant session. I don't know whether it matters who deleted and who re-creates the file (same person, host/non-host, etc.). A setup that did work for us:

1. Alice (host), Bob, and Carl start a session
2. Alice and Bob create continuous input in a file
3. Carl deletes the file
4. Bob re-creates the file

##### Expected behavior

The file is created normally (either empty or containing only the basic template) for all participants.

##### Actual behavior

The file is created for everybody. However, the file still contains the old content for Carl. (I don't know whether the issue happened for Carl due to the setup or just random chance.)

**Comment by tobous** [comment]
@srossbach You mentioned something about Saros/E retaining an old version of the editor content for too long:

> The problem is that we hold a reference for each user on the buffer of the editor. If these references are not cleared, i.e no editor close event is send, the content of the editor is still present on the host although the file that belong to the editor is already deleted.

Could this be related to the issue (even though the issue (also) happens for participants other than the host).

If not, do you have any other idea what might cause this issue?

I discovered this while testing the filter logic for #714, but I can't imagine how this issue could be connected. So, for now, I would like to continue the review process for #714 as it improves the logic, offers "new features" to the user (fixes file re-creation), and fixes the issue with the heartbeat dispatcher.

@srossbach Do you see any issue with this?

**Comment by tobous** [comment]
Also, for debugging purposes, I could provide the log files, but they are way to long to be really useful. On the first attempt, I was to slow with the file deletion, leading to a prolonged continuous modification by the other user and an overly large log file (~70mb, ~460.000 lines). The next test went more smoothly, but the log files are still around 60.000 lines each.

**Comment by tobous** [comment]
Ok, I tried it again and got it to work, but currently only **without** the latest fix for editor closed activities. So they might actually be the culprit. Will do some more testing. Here are the log files.
user `intellij_test_1` (host): [Saros_intellij_test_1@saros-con.imp.fu-berlin.de_2019-12-02_19-32-32.log](https://github.com/saros-project/saros/files/3912796/Saros_intellij_test_1%40saros-con.imp.fu-berlin.de_2019-12-02_19-32-32.log)

user `intellij_test_2`: 
[Saros_intellij_test_2@saros-con.imp.fu-berlin.de_2019-12-02_19-32-43.log](https://github.com/saros-project/saros/files/3912798/Saros_intellij_test_2%40saros-con.imp.fu-berlin.de_2019-12-02_19-32-43.log)

user `intellij_test_3`:

**Comment by srossbach** [comment]
The magic is done here https://github.com/saros-project/saros/blob/master/eclipse/src/saros/editor/RemoteWriteAccessManager.java

TLDR:

each connect increased the counter in the documentprovider for the document
each disconnect decreased the counter in the documentprovider for the document

if the counter reaches null then every change moved to /dev/null

if the counter is not 0 every IEditorManager#getContent will retrieve (depends on the documentprovider implementation, see various other issues, like XTEND plugin) some buffered content even if no editor is opened (and the storage file does not even exists)

**Comment by tobous** [comment]
Ok, so this should be fixed by the filter never filtering editor closed activities?

**Comment by tobous** [comment]
I tested again with the current version of #714 and could no longer reproduce this issue. So I would see this as resolved. @srossbach Any objections?

**Comment by srossbach** [comment]
Why is this issue tagged with Core ? Isn't it an Eclipse only issue ?

**Comment by tobous** [comment]
> Why is this issue tagged with Core ? Isn't it an Eclipse only issue ?

Probably. But when I created this issue, I didn't know that.

**Comment by tobous** [comment]
> I tested again with the current version of #714 and could no longer reproduce this issue. So I would see this as resolved.

As I am confident that the issue has been resolved, I don't see a reason to keep this issue open. In addition to no longer being reproducible in my testing setup, the [root cause mentioned above](https://github.com/saros-project/saros/issues/780#issuecomment-560719936) should have been resolved already:

- PR #752 ensures that all participants always close the editor for the deleted file, meaning corresponding editor closed activities should always be send if necessary.
- Furthermore, with the fix commit https://github.com/saros-project/saros/pull/714/commits/8ed15b5c3a1b26ae72fe91926784b91775355512 for PR #714, editor closed activities are also never filtered by the new resource activity filter.

As a result, the state of remote open editors should always be correctly updated on file deletion for all participants.

If no one objects to this decisions by next Monday (2019-12-09), I am going close this issue as resolved and move on with the review process for #714.

**Comment by srossbach** [comment]
I think it can be closed. Just as a reminder, what happens if there is an issue if a client has a malfunction and so the close editor event got swallowed due to an error.

**Comment by tobous** [comment]
> Just as a reminder, what happens if there is an issue if a client has a malfunction and so the close editor event got swallowed due to an error.

I have no idea how to handle that. (Well, actually, how about always dropping the reference when the matching file is deleted or created, now matter how many users we think still have it open. That increases the chance that the old editor reference is garbage collected before a new one is requested.)
But there are multiple similar issues that could happen if activities were dropped. We somewhat rely on the accuracy of the state reported by all participants.

But either way, this particular issue could also be solved by removing the read/write access feature (or at least adjusting it to be solely server-side-logic based, i.e. simply dropping activities on the server if the user doesn't have the right permissions).

But that can be discussed in more detail later on. I am currently in the process of starting my masters thesis dealing with an "architectural model" (i.e. how do we want Saros to look in its "final" state; no idea what form this will take/how precise it will be; the general gist is to provide some kind of "vision" to evaluate changes/new features against) and a set of functions for Saros under the specific aspect of making it IDE cross-compatible (i.e. which features have to go or have to be adjusted to allow for IDE-cross compatibility), creating a set of necessary steps and then starting to implement them.

And read/write access will be one of the topics (as I am not completely sold on the necessity/provided value and there are issues with a client side implementation for IntelliJ).

But that will still take some time as I am still in the initial research phase (and there is still some stuff I would like to get done before focusing on the thesis).
